runtime/server/psql.go

```diff
@@ -0,0 +1,214 @@
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/jackc/pgx/v5/pgtype"
+	wire "github.com/jeroenrinzema/psql-wire"
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime"
+	"github.com/rilldata/rill/runtime/server/auth"
+	"github.com/rilldata/rill/runtime/server/psql"
+)
+
+// psqlQueryHandler handles incoming queries on psql wire endpoint.
+// The metadata queries are redirected to ResolvePSQLQuery.
+// The metrics_sql queries are redirected to metrics_sql resolver.
+func (s *Server) psqlQueryHandler(ctx context.Context, query string) (wire.PreparedStatements, error) {
+	clientParams := wire.ClientParameters(ctx)
+	instanceID := clientParams[wire.ParamDatabase]
+
+	// The logic to identify metadata queries is somewhat hacky.
+	// Identify metadata queries based on whether query contains names of common postgres catalogs like pg_attribute, pg_catalog etc.
+	// Also consider queries that do not have a FROM clause as metadata queries since querying from metrics_view requires a FROM clause.
+	if strings.Contains(query, "pg_attribute") || strings.Contains(query, "pg_catalog") || strings.Contains(query, "pg_type") || strings.Contains(query, "pg_namespace") || !strings.Contains(strings.ToUpper(query), "FROM") {
+		data, schema, err := psql.ResolvePSQLQuery(ctx, &psql.PSQLQueryOpts{
+			SQL:            query,
+			Runtime:        s.runtime,
+			InstanceID:     instanceID,
+			UserAttributes: auth.GetClaims(ctx).Attributes(),
+			Priority:       1,
+			Logger:         s.logger,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		handle := func(ctx context.Context, writer wire.DataWriter, parameters []wire.Parameter) error {
+			for i := 0; i < len(data); i++ {
+				if err := writer.Row(data[i]); err != nil {
+					return fmt.Errorf("failed to write row: %w", err)
+				}
+			}
+			return writer.Complete("OK")
+		}
+		return wire.Prepared(wire.NewStatement(handle, wire.WithColumns(convertSchema(schema)))), nil
+	}
+
+	// todo handle normal SQL. Some ideas to handle that
+	// 1. Approach similar to BigQuery which has #legacySQL
+	// 2. We can parse for a -- @connector: xxx comment, and if found, use that connector directly instead of metrics SQL
+	res, err := s.runtime.Resolve(ctx, &runtime.ResolveOptions{
+		InstanceID:         instanceID,
+		Resolver:           "metrics_sql",
+		ResolverProperties: map[string]any{"sql": query},
+		Args:               map[string]any{"priority": 1},
+		UserAttributes:     auth.GetClaims(ctx).Attributes(),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	handle := func(ctx context.Context, writer wire.DataWriter, parameters []wire.Parameter) error {
```

Comment:
Is there *any* case where the `handle` function doesn't get invoked? E.g. what if the `ctx` is cancelled after `s.runtime.Resolve` completes, but before `wire.Prepared` runs? Because then the `res` would not be closed.

If yes, it would be safer to run the resolver inside the `handle` function. I think this might also be a problem in the `admin` service's psql handler.

Reply:
Yes it should be possible. Client connections can get killed anytime. 
Though this should be okay for runtime since it just gets results from cache and internally the results are closed within cache_resolver. 
Its a potential problem with admin.

The problem with running `Resolve` inside handler is that we need to set schema before client execute the query where this handler is called. Also the library doesn't allow to set schema later so we will need small changes there as well.
If we do that we end up in a situation similar to the issue and fix that we did here : https://github.com/rilldata/rill/pull/4978

I can see if I can use prepared statements between Admin and Runtime.